### PR TITLE
Add subscription tab

### PR DESCRIPTION
**Pull request by Sa1to**
It fixes [PLG: Add "Upgrade to Cody Pro" and "Current Usage" settings buttons#149]
(https://github.com/sourcegraph/jetbrains/issues/149)
We need to merge https://github.com/sourcegraph/cody/pull/2274 first
## Test plan
* Log in as Cody Free user
* Check if subscription tab available
* Check if you see correct tier
* Check if you see `Upgrade` and `Check Usage` buttons that are redirecting to `https://sourcegraph.com/cody/subscription` and `https://sourcegraph.com/cody/manage`
* Log in as Cody Pro user
* Check if subscription tab available
* Check if you see correct tier
* Check if you see only `Check Usage` buttons that are redirecting to `https://sourcegraph.com/cody/subscription` and `https://sourcegraph.com/cody/manage`

https://github.com/sourcegraph/jetbrains/assets/9321940/79ccf53d-0685-4b41-815e-511a0beb5f22

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/config/test.form
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/config/test.form:1]: I guess we can drop this file

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 1]
```diff
@@ -52,8 +53,9 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
   private val sendButton: JButton
   private var inProgressChat = CancellationToken()
   private val stopGeneratingButton =
-      JButton("Stop generating", IconUtil.desaturate(AllIcons.Actions.Suspend))
+    JButton("Stop generating", IconUtil.desaturate(AllIcons.Actions.Suspend))
   private val recipesPanel: JBPanelWithEmptyText
+  private val subscriptionPanel: JPanel
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: This file is already a mess. If there is a chance to add anything in a separate file in the first place we should go for it.

**Comment by vdavid** [comment]
Behavior-wise, 👍  from me as PLG team member. (I'll let Cody JetBrains teammates check the code.)

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 2]
```diff
@@ -111,6 +113,38 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
     addWelcomeMessage()
   }
 
+  @RequiresEdt
+  private fun addSubscriptionTab() {
+    tryRestartingAgentIfNotRunning(project)
+    val server = getServer(project)
+    if (server != null) {
+      val codyProFeatureFlag = server.evaluateFeatureFlag(GetFeatureFlag("CodyPro"))
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: tbd if we should limit to sourcegraph.com instance only

File: src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt [hunk 3]
```diff
@@ -226,7 +226,11 @@ class CodyAutocompleteManager {
                             com.sourcegraph.cody.vscode.Position(lineNumber, startPosition),
                             position)))
     notifyApplication(CodyAutocompleteStatus.AutocompleteInProgress)
-    val completions = server.autocompleteExecute(params)
+    //    runBlocking {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt]: nit: I guess the comment can be dropped

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 4]
```diff
@@ -111,6 +113,41 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
...
   }
 
+  @RequiresEdt
+  private fun addSubscriptionTab() {
+    val activeAccountType = CodyAuthenticationManager.instance.getActiveAccount(project)
+    if (activeAccountType != null && activeAccountType.isDotcomAccount()) {
+      tryRestartingAgentIfNotRunning(project)
+      val server = getServer(project)
+      if (server != null) {
+        val codyProFeatureFlag = server.evaluateFeatureFlag(GetFeatureFlag("CodyPro"))
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: we should depend on our flag now `cody-pro-jetbrains`

Comment (exigow) [reply src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: @mkondratek, can you elaborate and/or suggest a change? Szymon finished early today, so he won't be able to update this till tomorrow. 

Does this boil down to changing from `val codyProFeatureFlag = server.evaluateFeatureFlag(GetFeatureFlag("CodyPro"))` to `val codyProFeatureFlag = server.evaluateFeatureFlag(GetFeatureFlag("cody-pro-jetbrains"))`?

Comment (Sa1to) [reply src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: So should we rely **only** on our internal feature flag or should we have both (one form agent and our internal)?

**Review by exigow — CHANGES_REQUESTED** [review]
During tests, my IDE freezes and the last log is: `2023-12-11 18:27:13,812 [  37588]   WARN - #c.s.c.a.CodyAgent - No handler for request with method featureFlags/getFeatureFlag`

**Comment by Sa1to** [comment]
@exigow I have to rebase on master, it requires latest changes in agent 😅

**Review by exigow — APPROVED** [review]
Thanks for the fixes! :+1: 

Issue seems to be fixed with latest bump, but there is still an problem with the short IDE freezing (for less than a second but it's visible). Apparently, we have some code that happens in the AWT thread and it blocks the entire UI. 

We can ignore this problem for now.

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 5]
```diff
@@ -111,6 +113,43 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
...
+  private fun addSubscriptionTab() {
+    val activeAccountType = CodyAuthenticationManager.instance.getActiveAccount(project)
+    if (activeAccountType != null && activeAccountType.isDotcomAccount()) {
+      tryRestartingAgentIfNotRunning(project)
+      val server = getServer(project)
+      if (server != null) {
+        val codyProFeatureFlag = server.evaluateFeatureFlag(GetFeatureFlag("CodyProJetBrains"))
+        if (codyProFeatureFlag.get() != null &&
+            codyProFeatureFlag.get()!! &&
+            java.lang.Boolean.getBoolean("cody.isGa")) {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt]: As we decided during the meetings. Let's drop `cody.isGa` condition here. We will request ff exceptions for ones that want to test these features.

File: src/main/java/com/sourcegraph/cody/SubscriptionPanel.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  row {
+    if (!isCurrentUserPro) {
+      val upgradeButton =
+          button("Upgrade") { BrowserUtil.browse(ConfigUtil.DOTCOM_URL + "cody/subscription") }
+      upgradeButton.component.putClientProperty(DarculaButtonUI.DEFAULT_STYLE_KEY, true)
+    }
+    button("Check Usage") { BrowserUtil.browse(ConfigUtil.DOTCOM_URL + "cody/manage") }
+  }
+  if (!isCurrentUserPro) {
+    row { label("Please refresh IDE after upgrading to Cody Pro") }
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/SubscriptionPanel.kt]: pls use the copy from: https://github.com/sourcegraph/jetbrains/issues/178

**Review by mkondratek — CHANGES_REQUESTED** [review]
pls apply the two fixes I mentioned. 

There is one more nice to have - but we can deliver it later:
We could show the sub tab depending on the account type when it changes. For now when I change to the account that is eligible to see sub tab the tab does not appear. I assume you can add some logic there:
https://sourcegraph.com/github.com/sourcegraph/jetbrains/-/blob/src/main/kotlin/com/sourcegraph/cody/config/CodyPersistentAccountsHost.kt?L29%3A1-30%3A1

**Comment by mkondratek** [comment]
I created this: https://github.com/sourcegraph/jetbrains/issues/180

**Review by mkondratek — APPROVED** [review]
image


this is good 🚀

**Pull request merged**